### DISCUSSION TITLE
Does Node expose a way of hooking into node:perf_hooks?

### DISCUSSION THREAD

**Discussion by fatso83** [nodejs/help#4643] [Q&A] [answered]
I am maintaining the Sinon `fake-timers` library, and we have gotten a few requests to add support for `perf_hooks` (like  https://github.com/sinonjs/fake-timers/issues/462). @benjamingr mentioned that this might be possible, but he is currently occupied, so I am not sure what we have to work with. 

Does there exist some kind of hook that stubbing libraries such as ours can use to override the implementation? For instance, enabled through some command line argument to Node.

**Comment by softwarecreations** [comment, marked as answer]
Hi @fatso83
Node.js doesn't currently expose a public hook or API specifically for overriding or stubbing the high-resolution time implementation in `node:perf_hooks` (which is what powers performance.now() and similar). The underlying timer mechanism is tied closely to the V8 engine's `v8::Isolate::GetCurrentPlatform()` and libuv's high-resolution clock, and there's no built-in extension point for userland libraries to inject custom behavior there.

That said, there is an open RFC for adding configurable high-resolution time sources, which would make exactly this kind of overriding possible (e.g., via a hook or factory function). You can track it here: https://github.com/nodejs/tc39-perf-rfcs/pull/1. If it lands, it could integrate nicely with stubbing libraries like yours.In the meantime, a common workaround for testing scenarios is to monkey-patch the performance object at runtime in your test setup, like this:

```js
const originalNow = performance.now;
performance.now = () => {
  // Your stubbed logic here, e.g., return a fixed value or advance a mock clock
  return 12345.678;
};
```

This isn't as clean as a proper hook (and it won't affect the underlying perf hooks internals), but it works for most perf_hooks-related test faking needs. If you're aiming for deeper integration, you could also explore V8's `--allow-natives-syntax` flag for some introspection, though that's not recommended for production or stable tests.

If this doesn't quite fit your use case or if you have more details from the Sinon issue, feel free to share

**Reply by fatso83** [reply]
Thanks. We have always resorted to monkey-patching, but I initially thought we could not do that here, due to the exports being in a module. It took [a while](https://github.com/sinonjs/fake-timers/issues/462#issuecomment-2537286618) until I learned that while the namespace itself was immutable, much of the stuff we needed to patch were mutable fields on the exports. So no issue there in the end.
